### PR TITLE
MediatR - give it a look

### PR DESCRIPTION
**Pull request by dotNETSanta**
So, after reading slack and issues I'd want to ask whether we're heading into something like that?

Those models are mostly used as a dummy.

### CODE REVIEW

**Comment by silvuss** [comment]
@dotNETSanta How to treat this? A discussion or a pull request? In my opinion, a discussion is not a pull request, but this seems to be both.

**Comment by dotNETSanta** [comment]
@silvuss Well, I wanted to show some code to make "review"/"discussions" easier.

pinging @smknd

**Comment by silvuss** [comment]
OK, I will treat it as a pull request (to be reviewed).

File: Coyote/CoyoteNETCore.Shared/Thread.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace CoyoteNETCore.Shared
+{
+    public class Thread : Entity
+    {
+        public string Title { get; set; }
+
+        public string Body { get; set; }
```
[the remark this answers is not in the document]
Comment (dotNETSanta) [reply Coyote/CoyoteNETCore.Shared/Thread.cs]: Text, just text.

File: Coyote/CoyoteNETCore.Services/Thread/Command/CreateThreadCommandValidation.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+using MediatR;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CoyoteNETCore.Application.Thread.Command
+{
+    public class CreateThreadCommandValidation : AbstractValidator<CreateThreadCommand>
```
[the remark this answers is not in the document]
Comment (dotNETSanta) [reply Coyote/CoyoteNETCore.Services/Thread/Command/CreateThreadCommandValidation.cs]: How do you understand "flow"? HTTP Request  -> (middleware) -> controller -> validation -> command/query created from data sent to controller?

File: Coyote/CoyoteNETCore.Shared/Thread.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (silvuss) [reply Coyote/CoyoteNETCore.Shared/Thread.cs]: I think that the name `Body` is OK.

File: Coyote/CoyoteNETCore.Services/Thread/Command/CreateThreadCommand.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            Author = author ?? throw new Exception();
+        }
+
+        public string Body { get; set; }
+
+        public string Title { get; set; }
+
+        public User Author { get; set; }
+
+        public class Handler : IRequestHandler<CreateThreadCommand, (bool Success, string Result)>
```
[the remark this answers is not in the document]
Comment (dotNETSanta) [reply Coyote/CoyoteNETCore.Services/Thread/Command/CreateThreadCommand.cs]: Well, I used to keep 'em as a separated .cs, but then I've found that people tends to keep 'em together like that.

It seems to be quite reasonable, cuz it hides CTCommand's implementation. 

I think I'd be even better if we changed it to ``private class Handler.``

File: Coyote/CoyoteNETCore.Shared/Entity.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace CoyoteNETCore.Shared
+{
+    public class Entity
+    {
+        public Guid Id { get; protected set; } = Guid.NewGuid();
```
[the remark this answers is not in the document]
Comment (dotNETSanta) [reply Coyote/CoyoteNETCore.Shared/Entity.cs]: So, maybe we should remove an "Id" from ``Entity``?

**Comment by silvuss** [comment]
@dotNETSanta You closed it , but there are commits unmerged; is this expected?

**Comment by dotNETSanta** [comment]
This PR has been moved to: https://github.com/CoyoteNET/CoyoteNET-backend/pull/23

Because of troubles with getting checks.

File: Coyote/CoyoteNETCore.Shared/Thread.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (smknd) [reply Coyote/CoyoteNETCore.Shared/Thread.cs]: The thing is that post has body. Thread doesn't.

File: Coyote/CoyoteNETCore.Shared/Entity.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (smknd) [reply Coyote/CoyoteNETCore.Shared/Entity.cs]: Or make it `long` or `int`. Or make this class generic: `Entity<TIdType>`. 
Anyway using Guid for PK if you're not integrating data from a many sources is pointless.

File: Coyote/CoyoteNETCore.Shared/Thread.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (silvuss) [reply Coyote/CoyoteNETCore.Shared/Thread.cs]: @smknd That's how it should work. My bad. Of course, thread has a list of posts, and post have a body.

File: Coyote/CoyoteNETCore.Services/Thread/Command/CreateThreadCommand.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (smknd) [reply Coyote/CoyoteNETCore.Services/Thread/Command/CreateThreadCommand.cs]: If people tends to do something it's probably anti-pattern. :)

I have no experience with MediatR but I'm not sure if it should be nested. At first glance it looks like some kind of SRP violation, but of course I may be wrong. 
Maybe we can consider some kind of more generic approach consist on handlers which can directly use DTOs deserialized from HTTP request?

File: Coyote/CoyoteNETCore.Shared/Thread.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dotNETSanta) [reply Coyote/CoyoteNETCore.Shared/Thread.cs]: @smknd Depends on apporoach. If you threat thread text as a first post, then ok.

[the remark this answers is not in the document]
Comment (silvuss) [reply Coyote/CoyoteNETCore.Shared/Thread.cs]: @dotNETSanta Wouldn't it be less intuitive to implement? In my opinion, a thread = an empty container (with some additional attributes) for a list of posts. A post = a container for its title (text) and body (text).

**Pull request closed without merging**